cmd/crank/beta/trace/internal/resource/xrm/loader.go

```diff
@@ -0,0 +1,150 @@
+type resourceLoader interface {
+	loadResource(ctx context.Context, ref *v1.ObjectReference) *resource.Resource
+	getResourceChildrenRefs(r *resource.Resource) []v1.ObjectReference
+}
+
+// loader loads resources concurrently.
+type loader struct {
+	root         *resource.Resource // the root resource for which the tree is loaded
+	rl           resourceLoader     // the resource loader
+	resourceLock sync.Mutex         // lock when updating the children of any resource
+	processing   sync.WaitGroup     // "counter" to track requests in flight
+	ch           chan workItem      // processing channel
+	done         chan struct{}      // done channel, signaled when all resources are loaded
+}
+
+// newLoader creates a loader for the root resource.
+func newLoader(root *resource.Resource, rl resourceLoader, channelCapacity int) *loader {
+	l := &loader{
+		rl:   rl,
+		ch:   make(chan workItem, channelCapacity),
+		done: make(chan struct{}),
+		root: root,
+	}
+	return l
+}
+
+// load loads the full resource tree in a concurrent manner.
+func (l *loader) load(ctx context.Context, concurrency int) {
+	// make sure counters are incremented for root child refs before starting concurrent processing
+	refs := l.rl.getResourceChildrenRefs(l.root)
+	l.addRefs(l.root, refs)
+
+	// signal the done channel after all items are processed
+	go func() {
+		l.processing.Wait()
+		close(l.done)
+	}()
+
+	if concurrency < 1 {
+		concurrency = defaultConcurrency
+	}
+	var wg sync.WaitGroup
+	for range concurrency {
+		wg.Add(1)
+		// spin up a worker that processes items from the channel until the done channel is signaled.
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-l.done:
+					return
+				case item := <-l.ch:
+					l.processItem(ctx, item)
+				}
+			}
+		}()
+	}
+	wg.Wait()
+	// order of children loaded for resources is not deterministic because of concurrent processing.
+	// Sort children explicitly to make this so.
+	sortRefs(l.root)
+}
+
+func sortRefs(root *resource.Resource) {
+	for _, child := range root.Children {
+		sortRefs(child)
+	}
+	// this duplicates the sorting logic from internal/controller/apiextensions/composite/composition_functions.go
+	sort.Slice(root.Children, func(i, j int) bool {
+		l := root.Children[i].Unstructured
+		r := root.Children[j].Unstructured
+		return l.GetAPIVersion()+l.GetKind()+l.GetName() < r.GetAPIVersion()+r.GetKind()+r.GetName()
+	})
+}
+
+// addRefs adds work items to the queue.
+func (l *loader) addRefs(parent *resource.Resource, refs []v1.ObjectReference) {
+	// only perform work and spin up a goroutine if references are present.
+	if len(refs) == 0 {
+		return
+	}
+	// ensure counters are updated synchronously
+	l.processing.Add(len(refs))
+	// free up the current processing routine even if the channel would block.
+	go func() {
+		for _, ref := range refs {
+			l.ch <- workItem{
+				parent: parent,
+				child:  ref,
```

Comment:
Is this susceptible to the infamous go for loop variable gotcha? https://dev.to/kkentzo/the-golang-for-loop-gotcha-1n35

Looking a bit further, it appears this was fixed with go 1.22, which we appear to be using: https://github.com/crossplane/crossplane/blob/master/go.mod#L3 

Do we completely not have to worry about this anymore? or is there any further steps we would need to take? should we continue to be defensive about this regardless, out of habit/tradition? what do you think? 🤔 
https://go.dev/blog/loopvar-preview

Reply:
This code also works on Go 1.21 BTW, since there are no pointers or goroutines involved that use the loop variable outside its intended scope. Also as you rightly said Go 1.22 doesn't have this issue any more.

I can put a `ref := ref` in the loop if you think that is more consistent with the codebase, but I don't know that it is necessary.

Reply:
Ah yes, looking again more closely you are correct that this would work OK before go 1.22 also! A copy of the `ref` variable here is sent through the channel, so `ref` is never used outside of the scope of this loop 🙇‍♂️ 

No need to make a change here as far as I'm concerned!